pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -107,7 +104,55 @@ func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitG
 	}
 }
 
-func (w *HashAggPartialWorker) updatePartialResult(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, _ int) (err error) {
+func newPartialResults(rowNum int, finalConcurrency int) [][][]aggfuncs.PartialResult {
+	partialResults := make([][][]aggfuncs.PartialResult, finalConcurrency)
+	avgRowNum := rowNum / finalConcurrency
+	for i := 0; i < finalConcurrency; i++ {
+		partialResults[i] = make([][]aggfuncs.PartialResult, 0, avgRowNum)
+	}
+	return partialResults
+}
+
+// It will get partial results for all final workers according to the group by key
+func (w *HashAggPartialWorker) getPartialResultsForAllFinalWorker(_ *stmtctx.StatementContext, groupKey [][]byte, mapper []AggPartialResultMapper, finalConcurrency int) ([]int, [][][]aggfuncs.PartialResult) {
+	n := len(groupKey)
+	partialResults := newPartialResults(n, finalConcurrency)
+	allMemDelta := int64(0)
+	partialResultSize := w.getPartialResultSliceLenConsiderByteAlign()
+	finalWorkerIdxs := make([]int, n)
+	for i := 0; i < n; i++ {
+		finalWorkerIdx := int(murmur3.Sum32(groupKey[i])) % finalConcurrency
+		finalWorkerIdxs[i] = finalWorkerIdx
+		tmp, ok := mapper[finalWorkerIdx][string(groupKey[i])]
+
+		// This group by key has appeared before
+		if ok {
+			partialResults[finalWorkerIdx] = append(partialResults[finalWorkerIdx], tmp)
+			continue
+		}
+
+		// It's the first time that this group by key appeared, create it
+		partialResults[finalWorkerIdx] = append(partialResults[finalWorkerIdx], make([]aggfuncs.PartialResult, partialResultSize))
+		lastIdx := len(partialResults[finalWorkerIdx]) - 1
+		for j, af := range w.aggFuncs {
+			partialResult, memDelta := af.AllocPartialResult()
+			partialResults[finalWorkerIdx][lastIdx][j] = partialResult
+			allMemDelta += memDelta // the memory usage of PartialResult
+		}
+		allMemDelta += int64(partialResultSize * 8)
+		// Map will expand when count > bucketNum * loadFactor. The memory usage will double.
+		if len(mapper)+1 > (1<<w.BInMap)*hack.LoadFactorNum/hack.LoadFactorDen {
+			w.memTracker.Consume(hack.DefBucketMemoryUsageForMapStrToSlice * (1 << w.BInMap))
+			w.BInMap++
+		}
```

Comment:
The memory usage calculate is used for ONLY ONE map, not for many maps..

Reply:
> The memory usage calculate is used for ONLY ONE map, not for many maps..

okk, I will fix it.